fast64_internal/utility.py

```diff
@@ -523,6 +525,118 @@ def deleteIfFound(filePath, stringValue):
+
+def scale_mtx_from_vector(scale: mathutils.Vector):
+	scax = mathutils.Matrix.Scale(scale[0], 4, (1, 0, 0))
+	scay = mathutils.Matrix.Scale(scale[1], 4, (0, 1, 0))
+	scaz = mathutils.Matrix.Scale(scale[2], 4, (0, 0, 1))
+	return scax @ scay @ scaz
+
+def copy_object_and_apply(obj: bpy.types.Object, apply_scale = False, apply_modifiers = False):
+	if apply_scale or apply_modifiers:
+		# its a unique mesh, use object name
+		obj['instanced_mesh_name'] = obj.name
+		obj.original_name = obj.name
+		if apply_scale:
+			obj['original_mtx'] = translation_rotation_from_mtx(mathutils.Matrix(obj['original_mtx']))
+
+	obj_copy = obj.copy()
+	obj_copy.parent = None
+	# reset transformations
+	obj_copy.location = mathutils.Vector([0.0, 0.0, 0.0])
+	obj_copy.scale = mathutils.Vector([1.0, 1.0, 1.0])
+	obj_copy.rotation_quaternion = mathutils.Quaternion([1, 0, 0, 0])
+	obj_copy.data = obj_copy.data.copy()
+
+	if apply_modifiers:
+		# In order to correctly apply modifiers, we have to go through blender and add the object to the collection, then apply modifiers
+		prev_active = bpy.context.view_layer.objects.active
+		bpy.context.collection.objects.link(obj_copy)
+		obj_copy.select_set(True)
+		bpy.context.view_layer.objects.active = obj_copy
+		for modifier in obj_copy.modifiers:
+			attemptModifierApply(modifier)
+
+		bpy.context.view_layer.objects.active = prev_active
+
+	mtx = mathutils.Quaternion((1, 0, 0), math.radians(-90.0)).to_matrix().to_4x4()
+	if apply_scale:
+	 	mtx = mtx @ scale_mtx_from_vector(obj.scale)
+
+	obj_copy.data.transform(mtx)
+	# Flag used for finding these temp objects
+	obj_copy['temp_export'] = True
+
+	# Override for F3D culling bounds (used in addCullCommand)
+	bounds_mtx = scale_mtx_from_vector([1, -1, -1]) # reverse y/z
+	if apply_scale:
+		bounds_mtx *= scale_mtx_from_vector(obj.scale) # apply scale if needed
+	bounds_mtx = bounds_mtx @ mathutils.Quaternion((1, 0, 0), math.radians(-90.0)).to_matrix().to_4x4()
+	obj_copy['culling_bounds'] = rotate_bounds(obj_copy.bound_box, bounds_mtx)
+
+def store_original_meshes(add_warning: Callable[[str], None]):
+	'''
+		- Creates new objects at 0, 0, 0 with shared mesh
+		- Original mesh name is saved to each object
+	'''
+	instanced_meshes = set()
+	for obj in iter_from_active():
+		if obj.data is not None:
+			has_modifiers = len(obj.modifiers) != 0
+			has_uneven_scale = not obj_scale_is_unified(obj)
+			shares_mesh = obj.data.users > 1
+			if has_modifiers or has_uneven_scale:
+				if shares_mesh and has_modifiers:
+					add_warning(
+						f'Object "{obj.name}" cannot be instanced due to having modifiers so an extra displaylist will be created. Remove modifiers to allow instancing.')
+				if shares_mesh and has_uneven_scale:
+					add_warning(
+						f'Object "{obj.name}" cannot be instanced due to uneven object scaling and an extra displaylist will be created. Set all scale values to the same value to allow instancing.')
```

Comment:
I don't know why uniform scaling currently works but uneven scaling doesn't, but regardless I think uneven scaling could be implemented by multiplying the matrix on the rsp stack by whatever is needed (see `gsSPMatrix`)

Just saying I think it can be done, I'm not asking you to make it happen

Reply:
yeah thats definitely a good point, I'm avoiding it for now because I want this PR to be based in GeoLayout transformations (which are only uniform unfortunately). sometime in the future that'd be great shared functionality between oot/sm64